### PR TITLE
[quantization] Implement QuantGemma4ForConditionalGeneration PTQ wrapper

### PR DESCRIPTION
**Pull request by dvsav**
## What

This PR implements the full PTQ calibration support for `Gemma4ForConditionalGeneration` — the top-level multimodal model that wraps `Gemma4Model` (vision + text decoder) and adds an `lm_head` linear layer with logit softcapping. Previously this wrapper was a skeleton stub; now it has a complete calibration path with logit softcapping and fake-quantization observers.

## Why

The `QuantGemma4ForConditionalGeneration` wrapper existed only as a skeleton — its `forward()` returned raw logits without logit softcapping, and the module was commented out in the registry. This meant the top-level Gemma4 model could not be quantized. This PR completes the wrapper's calibration path so the prepare → calibrate → convert pipeline works, matching the behavior of the original `Gemma4ForConditionalGeneration.forward` including logit softcapping.

Per the [Support Gemma4](https://github.com/Samsung/TICO/issues/768) issue specification, the full Hugging Face Gemma4 forward is **not** exported as one NPU graph. Instead, the Llama static runtime pattern is followed, where `StaticGemma4Runtime` orchestrates the export of individual submodules. Therefore, this wrapper intentionally does **not** provide `as_export_module()` or `forward_export()` methods — the static runtime handles export at the submodule level.

## Key Design Decisions

1. **No export adapter** — Unlike `QuantGemma4Model`, this wrapper does not provide `as_export_module()` or `forward_export()`. The static runtime (`StaticGemma4Runtime`) exports individual submodules (text decoder layers, vision encoder, etc.) rather than the full model. Adding a top-level export adapter would contradict the issue specification and add unused code.

2. **Three observers for the softcapping path** — `obs_logit_softcapping_div` (after `logits / softcap`), `obs_logit_softcapping_tanh` (after `tanh`), and `obs_logits` (final logits). The `div` observer was added after discovering that the exported graph had a `div` node with missing quantization parameters. Each intermediate tensor in the softcapping chain gets its own observer so that every graph node carries qparam metadata.

3. **Extracted softcapping logic** — The div → fq → tanh → fq → mul → fq chain is encapsulated in a single `_apply_logit_softcapping()` helper method, keeping the `forward()` method clean and ensuring the softcapping logic is defined in one place.

4. **`logits_to_keep` tensor handling** — `forward()` matches the original's behavior: `int` → slice, `torch.Tensor` → advanced indexing, instead of silently falling through to `slice(None)` for tensor inputs.

5. **Calibration-only return type** — The wrapper returns raw logits (not `Gemma4CausalLMOutputWithPast`). This is sufficient for PTQ calibration where only logits matter for observer statistics. A TODO documents this for future HF compatibility.

## Changes

- **`tico/quantization/wrapq/wrappers/gemma4/quant_for_conditional_generation.py`** — Implemented `forward()` with logit softcapping and `logits_to_keep` tensor handling; added `_apply_logit_softcapping()` helper with three fake-quantization observers (`obs_logit_softcapping_div`, `obs_logit_softcapping_tanh`, `obs_logits`); updated `_all_observers()` to return the three new observers; removed the old skeleton `as_export_module()` that returned `Gemma4LMHeadExportAdapter`.
- **`tico/quantization/wrapq/wrappers/registry.py`** — Enabled the `quant_for_conditional_generation` module in `_CORE_MODULES` (uncommented).
- **`test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py`** — New file: unit tests covering wrapper type, forward finiteness, output shape, `logits_to_keep`, softcapping bounds, prepare-convert flow, and observer calibration.
- **`test/quantization/wrapq/wrappers/gemma4/test_quantize_for_conditional_generation.py`** — New file: smoke tests for the prepare-calibrate-convert flow, including FP parity, image-text calibration, `logits_to_keep`, and softcapping bounds.
- **`tico/quantization/wrapq/examples/gemma4/quantize_for_conditional_generation.py`** — New file: example script demonstrating the PTQ flow (prepare → calibrate → convert → PEIR comparison).
- **`tico/quantization/recipes/debug/wrapper_smoke/cases/gemma4.py`** — Added `Gemma4ForConditionalGenerationCase` smoke case and registered it in `GEMMA4_CASES`.

## Tests

All tests were run with `RUN_INTERNAL_TESTS=1`:
- `test_quant_for_conditional_generation.py` — Validates wrapper type, forward finiteness, output shape `(batch, seq_len, vocab_size)`, `logits_to_keep=1` slicing, softcapping bounds (`|logits| ≤ softcap`), full prepare-convert flow, and observer calibration after convert.
- `test_quantize_for_conditional_generation.py` — Validates FP parity before quantization, prepare-convert flow with text-only and image-text calibration data, `logits_to_keep` slicing, and softcapping bounds.
- Wrapper smoke runner — **PASS** with Mean |diff| = 0.000035, Max |diff| = 0.000540, PEIR = 0.000529%.

### Unit Tests

```
$ RUN_INTERNAL_TESTS=1 python -m pytest test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py -v
========================================================== test session starts ===========================================================
platform linux -- Python 3.10.12, pytest-9.0.3, pluggy-1.6.0 -- /home/d.savchenkov/myenv/bin/python
cachedir: .pytest_cache
rootdir: /home/d.savchenkov/TICO
configfile: pyproject.toml
plugins: anyio-4.13.0
collected 7 items                                                                                                                        

test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py::TestQuantGemma4ForConditionalGenerationSmoke::test_forward_output_shape PASSED [ 14%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py::TestQuantGemma4ForConditionalGenerationSmoke::test_logit_softcapping_applied PASSED [ 28%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py::TestQuantGemma4ForConditionalGenerationSmoke::test_logits_to_keep PASSED [ 42%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py::TestQuantGemma4ForConditionalGenerationSmoke::test_observers_calibrated PASSED [ 57%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py::TestQuantGemma4ForConditionalGenerationSmoke::test_prepare_convert_flow PASSED [ 71%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py::TestQuantGemma4ForConditionalGenerationSmoke::test_prepare_returns_correct_wrapper PASSED [ 85%]
test/quantization/wrapq/wrappers/gemma4/test_quant_for_conditional_generation.py::TestQuantGemma4ForConditionalGenerationSmoke::test_text_only_forward_is_finite PASSED [100%]
```

### Internal Tests

```
$ RUN_INTERNAL_TESTS=1 /home/d.savchenkov/myenv/bin/python -m pytest test/quantization/wrapq/wrappers/gemma4/test_quantize_for_conditional_generation.py -v
========================================================== test session starts ===========================================================
platform linux -- Python 3.10.12, pytest-9.0.3, pluggy-1.6.0 -- /home/d.savchenkov/myenv/bin/python
cachedir: .pytest_cache
rootdir: /home/d.savchenkov/TICO
configfile: pyproject.toml
plugins: anyio-4.13.0
collected 5 items                                                                                                                        

test/quantization/wrapq/wrappers/gemma4/test_quantize_for_conditional_generation.py::TestGemma4ForConditionalGenerationSmoke::test_logit_softcapping_bounded PASSED [ 20%]
test/quantization/wrapq/wrappers/gemma4/test_quantize_for_conditional_generation.py::TestGemma4ForConditionalGenerationSmoke::test_logits_to_keep PASSED [ 40%]
test/quantization/wrapq/wrappers/gemma4/test_quantize_for_conditional_generation.py::TestGemma4ForConditionalGenerationSmoke::test_no_quant_model_matches_reference PASSED [ 60%]
test/quantization/wrapq/wrappers/gemma4/test_quantize_for_conditional_generation.py::TestGemma4ForConditionalGenerationSmoke::test_prepare_convert_flow PASSED [ 80%]
test/quantization/wrapq/wrappers/gemma4/test_quantize_for_conditional_generation.py::TestGemma4ForConditionalGenerationSmoke::test_prepare_convert_flow_with_image PASSED [100%]

=========================================================== 5 passed in 44.41s ===========================================================
```

### Smoke Test

```
$ python -m tico.quantization.examples.inspect \
    --config tico/quantization/examples/configs/wrapper_smoke.yaml \
    --mode wrapper-smoke \
    --case gemma4_for_conditional_generation \
    --output-dir ./out/wrapper_smoke
┌───────────── Wrapper Smoke Summary ─────────────
│ Case             : gemma4_for_conditional_generation
│ Status           : PASS
│ Mean |diff|      : 0.000035
│ Max |diff|       : 0.000540
│ PEIR             : 0.000529
│ Shape match      : True
│ Quant finite     : True
└─────────────────────────────────────────────────
     ┌───────────────────────────────────────────┐
 0.56┤                                           │
     │                                       ••  │
     │                                     ••    │
 0.37┤                                   ••      │
     │                                 •••       │
     │                               •••         │
     │                             •••           │
 0.18┤                           •••             │
     │                         •••               │
     │                       •••                 │
-0.01┤                     •••                   │
     │                   •••                     │
     │                 •••                       │
     │               •••                         │
-0.19┤             •••                           │
     │           •••                             │
     │         •••                               │
-0.38┤       •••                                 │
     │     •••                                   │
     │   •••                                     │
     │  ••                                       │
-0.57┤                                           │
     └┬──────────┬─────────┬──────────┬─────────┬┘
    -0.57      -0.29     -0.01      0.28     0.56 
```

## Example Script

**`tico/quantization/wrapq/examples/gemma4/quantize_for_conditional_generation.py`** — Demonstrates the PTQ workflow:
1. Creates a tiny `Gemma4ForConditionalGeneration` with `final_logit_softcapping=30.0` (no download needed).
2. Prepares the model with `build_gemma4_e2b_ptq_config`.
3. Calibrates with 20 synthetic text-only samples.
4. Converts to a fake-quantized model.
5. Compares FP vs. quantized logits (prints PEIR and a plot).

Note: This script does not export to Circle format, as the full model export is handled by `StaticGemma4Runtime` at the submodule level, not by this wrapper.

```
$ python tico/quantization/wrapq/examples/gemma4/quantize_for_conditional_generation.py
Preparing model for quantization...
Calibrating (text-only)...
Converting to quantized model...

┌───────────── Quantization Error Summary ─────────────
│ FP output shape    : (1, 16, 256)
│ Quant output shape : (1, 16, 256)
│ Mean |diff|        : 0.000020
│ PEIR               : 0.010376 %
└──────────────────────────────────────────────────────
     ┌───────────────────────────────────────────┐
 0.50┤                                           │
     │                                       ••  │
     │                                     •     │
 0.32┤                                  •••      │
     │                               •••         │
     │                            ••••           │
 0.14┤                          •••              │
     │                        •••                │
-0.05┤                     •••                   │
     │                   •••                     │
     │                ••••                       │
-0.23┤              •••                          │
     │           •••                             │
     │          ••                               │
-0.41┤      •••                                  │
     │                                           │
     │  •                                        │
-0.59┤                                           │
     └┬──────────┬─────────┬──────────┬─────────┬┘
    -0.59      -0.32     -0.05      0.23     0.50 
```

### CODE REVIEW

File: tico/quantization/wrapq/wrappers/gemma4/quant_for_conditional_generation.py [hunk 1]
```diff
@@ -67,20 +75,117 @@ def forward(self, *args, logits_to_keep: int | torch.Tensor = 0, **kwargs):
...
+            per_layer_inputs=per_layer_inputs,
+            attention_masks=attention_masks,
+            position_embeddings=position_embeddings,
+        )
+
+        # Slice hidden states for logits_to_keep.
+        slice_indices = slice(-logits_to_keep, None) if logits_to_keep else slice(None)
+        logits = self.lm_head(hidden_states[:, slice_indices, :])
+
+        return self._apply_logit_softcapping(logits)
```
Comment (Torrero) [inline tico/quantization/wrapq/wrappers/gemma4/quant_for_conditional_generation.py]: if I understand correctly, `forward_export` should contain only `LM head` and `logit_softcapping` for NPU processing.

Comment (dvsav) [reply tico/quantization/wrapq/wrappers/gemma4/quant_for_conditional_generation.py]: Let me clarify my perspective on `forward_export`. Suppose we have a PTQ wrapper `QuantXXX`.
I assume that `QuantXXX.as_export_module()` method must return an object (called "export adapter" in TICO parlance) that implements computations **semantically indentical** to those performed by `QuantXXX.forward`, with one distinction: those computations must be export-friendly (they should not depend on dynamic tensor shapes and/or dynamic control flow).
`QuantXXX.forward_export` is intended to be called by the export adapter, so `QuantXXX.forward_export` must perform export-friendly computations semantically indentical to `QuantXXX.forward`.

Now, to the `QuantGemma4ForConditionalGeneration.forward_export`. If we let `forward_export` compute only a part of `forward` computations (only LM head + logit softcapping, without model), then `forward_export` will not be semantically equivalent to `forward`, right?

Please correct me if I'm wrong. I'd also appreciate if you clarified why you think that `forward_export` should contain only `LM head` and `logit_softcapping`.

Comment (Torrero) [reply tico/quantization/wrapq/wrappers/gemma4/quant_for_conditional_generation.py]: Ahh, I expressed myself incorrectly. Maybe `forward_export` should be excluded entirely for the `QuantGemma4ForConditionalGeneration` wrapper. As it is not assumed that the entire model will be exported to circle.

Comment (dvsav) [reply tico/quantization/wrapq/wrappers/gemma4/quant_for_conditional_generation.py]: 👍 I agree.
I've removed `QuantGemma4ForConditionalGeneration.as_export_module` and `QuantGemma4ForConditionalGeneration.forward_export` methods because:
1. [Support Gemma4](https://github.com/Samsung/TICO/issues/768) issue specification clearly states: "Do not try to export the full Hugging Face Gemma4 forward as one NPU graph. Instead, follow the Llama static runtime pattern".
2. `StaticGemma4Runtime` (defined in `tico/quantization/recipes/debug/static_gemma4_runtime.py`) doesn't use export adapters provided by higher-level modules: QuantGemma4Model, QuantGemma4ForConditionalGeneration, QuantGemma4ForCausalLM.

**Review by Torrero — APPROVED** [review]
LGTM

**Review by mhs4670go — APPROVED** [review]
LGTM
